[opaque_obj_v2] make_fx support

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #163284
* __->__ #165005

By wrapping the python objects with FakeScriptObject(FakeOpaqueQueue) we restrict users to do anything to this object. torch.compile support can be easily enabled by the rest of [this stack](https://github.com/pytorch/pytorch/pull/163936) and existing support for ScriptObjects.

One thing to note is that by default in functionalization we mark all ops that take in FakeScriptObjects as being effectful. Should this be the case for these custom ops that take in python objs?  


cc @ezyang @EikanWang @jgong5 @wenzhe-nrv

angelayi committed:
[opaque_obj_v2] make_fx support

[ghstack-poisoned]

angelayi committed:
Update on "[opaque_obj_v2] make_fx support"


By wrapping the python objects with FakeScriptObject(FakeOpaqueQueue) we restrict users to do anything to this object. torch.compile support can be easily enabled by the rest of [this stack](https://github.com/pytorch/pytorch/pull/163936) and existing support for ScriptObjects.

One thing to note is that by default in functionalization we mark all ops that take in FakeScriptObjects as being effectful. Should this be the case for these custom ops that take in python objs?  


cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

angelayi committed:
Update on "[opaque_obj_v2] make_fx support"


By wrapping the python objects with FakeScriptObject(FakeOpaqueQueue) we restrict users to do anything to this object. torch.compile support can be easily enabled by the rest of [this stack](https://github.com/pytorch/pytorch/pull/163936) and existing support for ScriptObjects.

One thing to note is that by default in functionalization we mark all ops that take in FakeScriptObjects as being effectful. Should this be the case for these custom ops that take in python objs?  


cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

angelayi committed:
Update on "[opaque_obj_v2] make_fx support"


By wrapping the python objects with FakeScriptObject(FakeOpaqueQueue) we restrict users to do anything to this object. torch.compile support can be easily enabled by the rest of [this stack](https://github.com/pytorch/pytorch/pull/163936) and existing support for ScriptObjects.

One thing to note is that by default in functionalization we mark all ops that take in FakeScriptObjects as being effectful. Should this be the case for these custom ops that take in python objs?  


cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

zou3519 commented on test/test_opaque_obj_v2.py:
```diff
@@ -53,6 +80,15 @@ def pop_impl(queue: OpaqueQueue) -> torch.Tensor:
+        def pop_impl_fake(q: OpaqueQueue) -> torch.Tensor:
+            # This is not accurate since the queue could have tensors that are
+            # not rank 1
+            ctx = torch._custom_op.impl.get_ctx()
```
isn't there a torch.library api for this

angelayi committed:
Update on "[opaque_obj_v2] make_fx support"


By wrapping the python objects with FakeScriptObject(FakeOpaqueQueue) we restrict users to do anything to this object. torch.compile support can be easily enabled by the rest of [this stack](https://github.com/pytorch/pytorch/pull/163936) and existing support for ScriptObjects.

One thing to note is that by default in functionalization we mark all ops that take in FakeScriptObjects as being effectful. Should this be the case for these custom ops that take in python objs?  


cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

angelayi commented:
@pytorchbot merge -i
